File: backend/routers/onboarding.py
"""T-V3-B-29 / F-027: Onboarding flow REST API.

OpenAPI spec: docs/api-design/2026-05-16_v3/openapi.yaml
  - GET   /api/me/onboarding         → 現在の onboarding 状態 (state, current_step, completed)
  - POST  /api/me/onboarding/advance → step を進めて next_step を返す
  - POST  /api/me/onboarding/skip    → optional step を skip / skipped_at を返す

AC マッピング (audit: docs/audit/2026-05-16_v3/T-V3-B-29.md):
  AC-F1 EVENT-DRIVEN  : POST /advance with valid step → persist + return next_step
  AC-F2 UNWANTED      : POST /skip for required step  → 409
  AC-F3 EVENT-DRIVEN  : GET  /onboarding              → 2xx with state
  AC-F4 UNWANTED      : GET  unauthorized             → 401
  AC-F5 UNWANTED      : GET  invalid body             → 422
  AC-F6 EVENT-DRIVEN  : POST /advance with valid auth → 2xx with next_step
  AC-F7 UNWANTED      : POST /advance unauthorized    → 401
  AC-F8 UNWANTED      : POST /advance invalid body    → 422
  AC-F9 EVENT-DRIVEN  : POST /skip with valid auth    → 2xx with skipped_at
  AC-F10 UNWANTED     : POST /skip unauthorized       → 401
"""
from __future__ import annotations

from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from services import onboarding as svc
from services.auth_middleware import require_user

router = APIRouter(prefix="/api/me/onboarding", tags=["onboarding"])


# ── Pydantic schemas ────────────────────────────────────

class AdvanceRequest(BaseModel):
    step: str = Field(..., min_length=1, max_length=100, description="完了マークする step ID")
    payload: dict[str, Any] = Field(default_factory=dict, description="その step の入力データ")


class SkipRequest(BaseModel):
    step: Optional[str] = Field(default=None, max_length=100, description="skip 対象 step (未指定なら current_step)")
    reason: Optional[str] = Field(default=None, max_length=500)


class OnboardingStateResponse(BaseModel):
    state: dict[str, Any]
    current_step: str
    completed: bool
    completed_steps: list[str]
    skipped_steps: list[str]
    completed_at: Optional[str] = None
    skipped_at: Optional[str] = None


class AdvanceResponse(BaseModel):
    next_step: Optional[str]
    completed: bool
    current_step: str


class SkipResponse(BaseModel):
    skipped_at: str
    next_step: Optional[str]
    completed: bool


# ── error helper ─────────────────────────────────────────

def _err(code: str, message: str, *, status_code: int) -> HTTPException:
    return HTTPException(status_code=status_code, detail={"code": code, "message": message})


def _map_service_error(e: svc.OnboardingError) -> HTTPException:
    if isinstance(e, svc.RequiredStepSkipError):
        return _err("required_step_skip", str(e), status_code=409)
    if isinstance(e, svc.StepOutOfOrderError):
        return _err("step_out_of_order", str(e), status_code=409)
    if isinstance(e, svc.UnknownStepError):
        return _err("unknown_step", str(e), status_code=422)
    return _err("onboarding_error", str(e), status_code=400)


def _user_id(user: dict) -> str:
    return str(user.get("sub") or user.get("id") or "")


# ── endpoints ─────────────────────────────────────────────

@router.get("", response_model=OnboardingStateResponse)
async def get_me_onboarding(user: dict = Depends(require_user)) -> dict:
    """AC-F3: GET /api/me/onboarding — 現在の onboarding 状態を返す.
    AC-F4: require_user が unauthorized → 401 を投げる.
    """
    uid = _user_id(user)
    state = await svc.get_state(uid)
    return {
        "state":           state["state"],
        "current_step":    state["current_step"],
        "completed":       state["completed"],
        "completed_steps": state["completed_steps"],
        "skipped_steps":   state["skipped_steps"],
        "completed_at":    state["completed_at"],
        "skipped_at":      state["skipped_at"],
    }


@router.post("/advance", response_model=AdvanceResponse, status_code=201)
async def post_me_onboarding_advance(
    body: AdvanceRequest,
    user: dict = Depends(require_user),
) -> dict:
    """AC-F1 / AC-F6: POST /advance — step を進めて next_step を返す.
    AC-F7: unauthorized → 401 (require_user).
    AC-F8: invalid body → 422 (Pydantic).
    """
    uid = _user_id(user)
    try:
        return await svc.advance(uid, body.step, body.payload)
    except svc.OnboardingError as e:
        raise _map_service_error(e)


@router.post("/skip", response_model=SkipResponse, status_code=201)
async def post_me_onboarding_skip(
    body: SkipRequest,
    user: dict = Depends(require_user),
) -> dict:
    """AC-F2 / AC-F9: POST /skip — optional step を skip / skipped_at 返す.
    AC-F2: required step → 409 (RequiredStepSkipError).
    AC-F10: unauthorized → 401 (require_user).
    """
    uid = _user_id(user)
    try:
        return await svc.skip(uid, body.step, body.reason)
    except svc.OnboardingError as e:
        raise _map_service_error(e)

File: frontend/src/api/onboarding.ts
/**
 * T-V3-C-39 / S-048 — Typed client for /api/me/onboarding (Feature F-027).
 *
 * Backend contracts (T-V3-B-29 ONBOARDING 実装済 — backend/routers/onboarding.py):
 *   GET   /api/me/onboarding          — get_me_onboarding
 *   POST  /api/me/onboarding/advance  — post_me_onboarding_advance
 *   POST  /api/me/onboarding/skip     — post_me_onboarding_skip
 *
 * OpenAPI: docs/api-design/2026-05-16_v3/openapi.yaml#/paths/~1api~1me~1onboarding
 *
 * EARS AC mapping (T-V3-C-39):
 *   AC-F1: 401 → page redirect to /login (S-001) handled by the page via
 *          {@link OnboardingApiError.status} === 401
 *   AC-F2: GET state during skeleton phase → {@link getOnboardingState}
 *
 * Non-technical error toast envelope follows the project-wide FastAPI contract:
 *   { detail: { code: string; message: string } }
 */

import { env } from "@/env";

export const ONBOARDING_ENDPOINT = "/api/me/onboarding";
export const ONBOARDING_ADVANCE_ENDPOINT = "/api/me/onboarding/advance";
export const ONBOARDING_SKIP_ENDPOINT = "/api/me/onboarding/skip";

// --------------------------------------------------------------------------
// Types (mirror backend Pydantic models — backend/routers/onboarding.py)
// --------------------------------------------------------------------------

/** GET /api/me/onboarding response shape. */
export interface OnboardingStateResponse {
  /** Opaque state descriptor (welcome|setup|ai_intro|done). */
  state: string;
  /** Current onboarding step id (e.g. "welcome"). */
  current_step: string;
  /** True once all required steps have been advanced or skipped. */
  completed: boolean;
}

/** POST /api/me/onboarding/advance request shape. */
export interface AdvanceRequest {
  /** Step id being completed (e.g. "welcome"). */
  step: string;
  /** Step-scoped payload (free-form per backend service contract). */
  payload?: Record<string, unknown>;
}

/** POST /api/me/onboarding/advance response shape. */
export interface AdvanceResponse {
  /** Next step id (null when flow completes). */
  next_step: string | null;
  /** True iff the advance call completed the full flow. */
  completed: boolean;
  /** Echo of the step now active after the advance (welcome→setup, etc.). */
  current_step?: string;
}

/** POST /api/me/onboarding/skip request shape. */
export interface SkipRequest {
  /** Optional step id to skip (defaults to current_step server-side). */
  step?: string;
}

/** POST /api/me/onboarding/skip response shape. */
export interface SkipResponse {
  /** ISO 8601 timestamp at which the skip was recorded. */
  skipped_at: string;
  /** Next step id, or null when the flow short-circuits to done. */
  next_step?: string | null;
  /** Whether the entire flow is now considered complete. */
  completed?: boolean;
}

// --------------------------------------------------------------------------
// Error class (AC-F1: 401 surfaces non-leaky message + lets caller redirect)
// --------------------------------------------------------------------------

interface BackendErrorEnvelope {
  detail?: { code?: string; message?: string } | string;
}

/**
 * Thrown for any non-2xx response from the /api/me/onboarding endpoints.
 *
 * `.toUserMessage()` returns a non-technical, end-user friendly message that
 * references the failing endpoint and never includes raw server stack traces,
 * backend exception class names, or SQL details.
 */
export class OnboardingApiError extends Error {
  readonly code: string;
  readonly status: number;
  readonly endpoint: string;

  constructor(code: string, message: string, status: number, endpoint: string) {
    super(message);
    this.name = "OnboardingApiError";
    this.code = code;
    this.status = status;
    this.endpoint = endpoint;
  }

  toUserMessage(): string {
    const friendly = friendlyMessageForStatus(this.status);
    return `${friendly} (${this.endpoint})`;
  }
}

const USER_MESSAGES: Record<number | "default", string> = {
  400: "入力内容を確認してください",
  401: "サインインが必要です",
  403: "この操作を実行する権限がありません",
  404: "オンボーディング状態が見つかりませんでした",
  409: "ステップの順序が一致しません",
  422: "入力フォーマットが正しくありません",
  429: "リクエストが多すぎます。しばらく待って再試行してください",
  500: "サーバーエラーが発生しました。時間をおいて再試行してください",
  default: "通信に失敗しました",
};

function friendlyMessageForStatus(status: number): string {
  return USER_MESSAGES[status] ?? USER_MESSAGES.default;
}

// --------------------------------------------------------------------------
// Internal HTTP helper
// --------------------------------------------------------------------------

export interface OnboardingRequestOptions {
  apiBase?: string;
  signal?: AbortSignal;
  /** Bearer token (Authorization: Bearer ...) — authenticated role required. */
  authToken?: string | null;
  /** Test seam — defaults to global fetch. */
  fetchImpl?: typeof fetch;
}

function resolveApiBase(opts: OnboardingRequestOptions): string {
  if (opts.apiBase) return opts.apiBase;
  const fromEnv = env.NEXT_PUBLIC_API_URL;
  return (fromEnv ?? "http://localhost:8001").replace(/\/$/, "");
}

async function parseError(
  response: Response,
  endpoint: string,
): Promise<OnboardingApiError> {
  let code = "UNKNOWN";
  let message = response.statusText || "request failed";
  try {
    const payload = (await response.json()) as BackendErrorEnvelope;
    if (payload && typeof payload.detail === "object" && payload.detail) {
      if (typeof payload.detail.code === "string") code = payload.detail.code;
      if (typeof payload.detail.message === "string") {
        message = payload.detail.message;
      }
    } else if (typeof payload?.detail === "string") {
      message = payload.detail;
    }
  } catch {
    // Non-JSON body — keep the synthesised message. Never leak raw body.
  }
  return new OnboardingApiError(code, message, response.status, endpoint);
}

async function request<TOut>(
  endpoint: string,
  init: RequestInit,
  opts: OnboardingRequestOptions,
): Promise<TOut> {
  const base = resolveApiBase(opts);
  const fetchImpl = opts.fetchImpl ?? fetch;
  const url = `${base}${endpoint}`;

  const headers: Record<string, string> = {
    Accept: "application/json",
    ...(init.headers as Record<string, string> | undefined),
  };
  if (init.body && !headers["Content-Type"]) {
    headers["Content-Type"] = "application/json";
  }
  if (opts.authToken) {
    headers.Authorization = `Bearer ${opts.authToken}`;
  }

  let response: Response;
  try {
    response = await fetchImpl(url, {
      ...init,
      headers,
      signal: opts.signal,
    });
  } catch (err) {
    if ((err as { name?: string }).name === "AbortError") throw err;
    throw new OnboardingApiError(
      "NETWORK_ERROR",
      "network error",
      0,
      endpoint,
    );
  }

  if (!response.ok) {
    throw await parseError(response, endpoint);
  }
  if (response.status === 204) return {} as TOut;
  return (await response.json()) as TOut;
}

// --------------------------------------------------------------------------
// Public API
// --------------------------------------------------------------------------

/** GET /api/me/onboarding — current onboarding flow state. */
export function getOnboardingState(
  opts: OnboardingRequestOptions = {},
): Promise<OnboardingStateResponse> {
  return request<OnboardingStateResponse>(
    ONBOARDING_ENDPOINT,
    { method: "GET" },
    opts,
  );
}

/** POST /api/me/onboarding/advance — mark a step as completed and pull the next. */
export function advanceOnboarding(
  body: AdvanceRequest,
  opts: OnboardingRequestOptions = {},
): Promise<AdvanceResponse> {
  return request<AdvanceResponse>(
    ONBOARDING_ADVANCE_ENDPOINT,
    { method: "POST", body: JSON.stringify({ payload: {}, ...body }) },
    opts,
  );
}

/** POST /api/me/onboarding/skip — skip the current (or named) step. */
export function skipOnboarding(
  body: SkipRequest = {},
  opts: OnboardingRequestOptions = {},
): Promise<SkipResponse> {
  return request<SkipResponse>(
    ONBOARDING_SKIP_ENDPOINT,
    { method: "POST", body: JSON.stringify(body) },
    opts,
  );
}
